File: src/graphVerification.ts
export interface GraphVerificationFile {
  path: string;
  length: number;
  digest: string;
}

export interface GraphVerificationManifest {
  schemaVersion: 1;
  tool: "tine-graph-bytes";
  algorithm: "sha256";
  complete: boolean;
  generatedAtUnixMs: number;
  files: GraphVerificationFile[];
  aggregateDigest?: string;
  errors: Array<{ path?: string; detail: string }>;
}

export interface GraphVerificationComparison {
  matches: boolean;
  incomplete: boolean;
  localOnly: string[];
  otherOnly: string[];
  changed: string[];
}

const SHA256 = /^[0-9a-f]{64}$/;

function isSafeRelativePath(path: string): boolean {
  if (!path || path.startsWith("/") || path.includes("\\") || path.includes("\0")) return false;
  return path.split("/").every((part) => part !== "" && part !== "." && part !== "..");
}

export function parseGraphVerificationManifest(text: string): GraphVerificationManifest {
  const value: unknown = JSON.parse(text);
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error("The comparison report is not a JSON object.");
  }
  const manifest = value as Record<string, unknown>;
  if (manifest.schemaVersion !== 1 || manifest.tool !== "tine-graph-bytes" || manifest.algorithm !== "sha256") {
    throw new Error("The comparison report has an unsupported format.");
  }
  if (typeof manifest.complete !== "boolean" || !Number.isSafeInteger(manifest.generatedAtUnixMs)) {
    throw new Error("The comparison report is missing required metadata.");
  }
  if (!Array.isArray(manifest.files) || !Array.isArray(manifest.errors)) {
    throw new Error("The comparison report has invalid file or error entries.");
  }

  const files = manifest.files.map((entry): GraphVerificationFile => {
    if (!entry || typeof entry !== "object" || Array.isArray(entry)) {
      throw new Error("The comparison report has an invalid file entry.");
    }
    const file = entry as Record<string, unknown>;
    if (typeof file.path !== "string" || !isSafeRelativePath(file.path)) {
      throw new Error("The comparison report contains an unsafe file path.");
    }
    if (!Number.isSafeInteger(file.length) || (file.length as number) < 0) {
      throw new Error(`The comparison report has an invalid length for ${file.path}.`);
    }
    if (typeof file.digest !== "string" || !SHA256.test(file.digest)) {
      throw new Error(`The comparison report has an invalid digest for ${file.path}.`);
    }
    return { path: file.path, length: file.length as number, digest: file.digest };
  });
  files.sort((left, right) => left.path.localeCompare(right.path));
  if (files.some((file, index) => index > 0 && file.path === files[index - 1].path)) {
    throw new Error("The comparison report contains duplicate file paths.");
  }
  if (manifest.aggregateDigest !== undefined &&
      (typeof manifest.aggregateDigest !== "string" || !SHA256.test(manifest.aggregateDigest))) {
    throw new Error("The comparison report has an invalid aggregate digest.");
  }
  for (const entry of manifest.errors) {
    if (!entry || typeof entry !== "object" || Array.isArray(entry) ||
        typeof (entry as Record<string, unknown>).detail !== "string") {
      throw new Error("The comparison report has an invalid error entry.");
    }
  }
  return {
    schemaVersion: 1,
    tool: "tine-graph-bytes",
    algorithm: "sha256",
    complete: manifest.complete,
    generatedAtUnixMs: manifest.generatedAtUnixMs as number,
    files,
    aggregateDigest: manifest.aggregateDigest as string | undefined,
    errors: manifest.errors as Array<{ path?: string; detail: string }>,
  };
}

export function compareGraphVerificationManifests(
  local: GraphVerificationManifest,
  other: GraphVerificationManifest,
): GraphVerificationComparison {
  if (!local.complete || !other.complete) {
    return { matches: false, incomplete: true, localOnly: [], otherOnly: [], changed: [] };
  }
  const localFiles = new Map(local.files.map((file) => [file.path, file]));
  const otherFiles = new Map(other.files.map((file) => [file.path, file]));
  const localOnly = [...localFiles.keys()].filter((path) => !otherFiles.has(path)).sort();
  const otherOnly = [...otherFiles.keys()].filter((path) => !localFiles.has(path)).sort();
  const changed = [...localFiles.keys()]
    .filter((path) => {
      const remote = otherFiles.get(path);
      const current = localFiles.get(path)!;
      return remote !== undefined && (remote.length !== current.length || remote.digest !== current.digest);
    })
    .sort();
  return {
    matches: localOnly.length === 0 && otherOnly.length === 0 && changed.length === 0,
    incomplete: false,
    localOnly,
    otherOnly,
    changed,
  };
}

File: src-tauri/src/graph_verification.rs
use crate::state::{slot_for_context, GraphContext};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::{hash_map::Entry, HashMap};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Mutex, OnceLock};
use std::time::{SystemTime, UNIX_EPOCH};
use tauri::{AppHandle, Emitter};
use tine_core::model::{Graph, GraphTextSourceDigest};

const MANIFEST_TOOL: &str = "tine-graph-bytes";
const MANIFEST_ALGORITHM: &str = "sha256";

static VERIFICATIONS: OnceLock<Mutex<HashMap<String, Arc<AtomicBool>>>> = OnceLock::new();

pub(crate) fn graph_verification_cancelled_error() -> crate::command_error::CommandError {
    crate::command_error::CommandError::tagged("operation-cancelled", None::<String>, None)
}

fn verifications() -> &'static Mutex<HashMap<String, Arc<AtomicBool>>> {
    VERIFICATIONS.get_or_init(|| Mutex::new(HashMap::new()))
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct GraphVerificationProgress {
    operation_id: String,
    processed: usize,
    total: usize,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct GraphVerificationError {
    #[serde(skip_serializing_if = "Option::is_none")]
    path: Option<String>,
    detail: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct GraphVerificationManifest {
    schema_version: u32,
    tool: String,
    algorithm: String,
    complete: bool,
    generated_at_unix_ms: u128,
    files: Vec<GraphTextSourceDigest>,
    #[serde(skip_serializing_if = "Option::is_none")]
    aggregate_digest: Option<String>,
    errors: Vec<GraphVerificationError>,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct GraphVerificationReport {
    text: String,
    suggested_file_name: String,
    total_files: usize,
    total_bytes: u64,
    aggregate_digest: Option<String>,
    complete: bool,
}

struct VerificationRegistration {
    operation_id: String,
}

impl Drop for VerificationRegistration {
    fn drop(&mut self) {
        if let Ok(mut jobs) = verifications().lock() {
            jobs.remove(&self.operation_id);
        }
    }
}

fn aggregate_digest(files: &[GraphTextSourceDigest]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(b"tine-graph-bytes-v1\0");
    for file in files {
        hasher.update((file.path.len() as u64).to_be_bytes());
        hasher.update(file.path.as_bytes());
        hasher.update(file.length.to_be_bytes());
        hasher.update(file.digest.as_bytes());
    }
    format!("{:x}", hasher.finalize())
}

fn build_manifest(
    graph: &Graph,
    app: &AppHandle,
    operation_id: &str,
    cancelled: &AtomicBool,
) -> Result<GraphVerificationManifest, crate::command_error::CommandError> {
    let mut errors = Vec::new();
    let mut paths = match graph.graph_text_source_paths() {
        Ok(paths) => paths,
        Err(error) => {
            errors.push(GraphVerificationError {
                path: None,
                detail: format!("source inventory failed: {error}"),
            });
            Vec::new()
        }
    };
    paths.sort();
    let total = paths.len();
    let mut files = Vec::with_capacity(total);
    for (index, path) in paths.iter().enumerate() {
        if cancelled.load(Ordering::Acquire) {
            return Err(graph_verification_cancelled_error());
        }
        match graph.digest_graph_text_source(path, cancelled) {
            Ok(file) => files.push(file),
            Err(error) if error.kind() == std::io::ErrorKind::Interrupted => {
                return Err(graph_verification_cancelled_error());
            }
            Err(error) => errors.push(GraphVerificationError {
                path: Some(path.clone()),
                detail: error.to_string(),
            }),
        }
        let _ = app.emit(
            "graph-verification-progress",
            GraphVerificationProgress {
                operation_id: operation_id.to_owned(),
                processed: index + 1,
                total,
            },
        );
    }
    files.sort_by(|left, right| left.path.cmp(&right.path));

    match graph.graph_text_source_paths() {
        Ok(mut after) => {
            after.sort();
            if after != paths {
                errors.push(GraphVerificationError {
                    path: None,
                    detail: "source inventory changed while verification was running".into(),
                });
            }
        }
        Err(error) => errors.push(GraphVerificationError {
            path: None,
            detail: format!("final source inventory failed: {error}"),
        }),
    }

    let complete = errors.is_empty() && files.len() == paths.len();
    Ok(GraphVerificationManifest {
        schema_version: 1,
        tool: MANIFEST_TOOL.into(),
        algorithm: MANIFEST_ALGORITHM.into(),
        complete,
        generated_at_unix_ms: SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_millis(),
        aggregate_digest: complete.then(|| aggregate_digest(&files)),
        files,
        errors,
    })
}

#[tauri::command]
pub(crate) async fn create_graph_verification(
    state: GraphContext<'_>,
    app: AppHandle,
    operation_id: String,
) -> Result<GraphVerificationReport, crate::command_error::CommandError> {
    if operation_id.is_empty() || operation_id.len() > 128 {
        return Err(crate::command_error::CommandError::prose(
            "invalid graph verification operation id",
        ));
    }
    let slot = slot_for_context(&state).map_err(crate::command_error::CommandError::from)?;
    drop(state);
    let cancelled = Arc::new(AtomicBool::new(false));
    {
        let mut jobs = verifications().lock().map_err(|_| {
            crate::command_error::CommandError::prose("graph verification registry is unavailable")
        })?;
        match jobs.entry(operation_id.clone()) {
            Entry::Vacant(entry) => {
                entry.insert(Arc::clone(&cancelled));
            }
            Entry::Occupied(_) => {
                return Err(crate::command_error::CommandError::prose(
                    "graph verification operation id is already active",
                ));
            }
        }
    }
    let registration = VerificationRegistration {
        operation_id: operation_id.clone(),
    };
    tauri::async_runtime::spawn_blocking(move || {
        let _registration = registration;
        let manifest = slot
            .with_filesystem_graph(|graph| {
                build_manifest(graph, &app, &operation_id, &cancelled)
                    .map_err(crate::command_error::CommandError::prose)
            })
            .map_err(crate::command_error::CommandError::from)?;
        let total_bytes = manifest
            .files
            .iter()
            .try_fold(0_u64, |sum, file| sum.checked_add(file.length))
            .ok_or_else(|| {
                crate::command_error::CommandError::prose("graph verification byte count overflow")
            })?;
        let text = serde_json::to_string_pretty(&manifest).map_err(|error| {
            crate::command_error::CommandError::json(format!(
                "graph verification report could not be encoded: {error}"
            ))
        })?;
        Ok(GraphVerificationReport {
            text,
            suggested_file_name: format!(
                "tine-graph-verification-{}.json",
                manifest.generated_at_unix_ms
            ),
            total_files: manifest.files.len(),
            total_bytes,
            aggregate_digest: manifest.aggregate_digest,
            complete: manifest.complete,
        })
    })
    .await
    .map_err(|error| {
        crate::command_error::CommandError::graph_verification(format!(
            "graph verification task failed: {error}"
        ))
    })?
}

#[tauri::command]
pub(crate) fn cancel_graph_verification(
    operation_id: String,
) -> Result<(), crate::command_error::CommandError> {
    let jobs = verifications().lock().map_err(|_| {
        crate::command_error::CommandError::prose("graph verification registry is unavailable")
    })?;
    if let Some(cancelled) = jobs.get(&operation_id) {
        cancelled.store(true, Ordering::Release);
    }
    Ok(())
}

#[tauri::command]
pub(crate) async fn save_graph_verification_report(
    app: AppHandle,
    text: String,
) -> Result<bool, crate::command_error::CommandError> {
    let manifest: GraphVerificationManifest = serde_json::from_str(&text).map_err(|error| {
        crate::command_error::CommandError::json(format!(
            "graph verification report is invalid: {error}"
        ))
    })?;
    if manifest.schema_version != 1
        || manifest.tool != MANIFEST_TOOL
        || manifest.algorithm != MANIFEST_ALGORITHM
    {
        return Err(crate::command_error::CommandError::prose(
            "graph verification report has an unsupported format",
        ));
    }
    #[cfg(desktop)]
    {
        use tauri_plugin_dialog::DialogExt as _;
        let suggested = format!(
            "tine-graph-verification-{}.json",
            manifest.generated_at_unix_ms
        );
        let chosen = tauri::async_runtime::spawn_blocking(move || {
            app.dialog()
                .file()
                .set_file_name(suggested)
                .add_filter("JSON", &["json"])
                .blocking_save_file()
        })
        .await
        .map_err(|error| {
            crate::command_error::CommandError::graph_verification(format!(
                "graph verification save dialog failed: {error}"
            ))
        })?;
        let Some(chosen) = chosen else {
            return Ok(false);
        };
        let path = chosen.into_path().map_err(|_| {
            crate::command_error::CommandError::prose(
                "graph verification destination is not a local file",
            )
        })?;
        tine_core::model::atomic_write(&path, text.as_bytes()).map_err(|error| {
            crate::command_error::CommandError::graph_verification(format!(
                "graph verification report could not be saved: {error}"
            ))
        })?;
        Ok(true)
    }
    #[cfg(not(desktop))]
    {
        let _ = (app, manifest, text);
        Err(crate::command_error::CommandError::prose(
            "Save report is available on desktop; use Copy report on this device.",
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn aggregate_is_ordered_and_content_sensitive() {
        let first = vec![GraphTextSourceDigest {
            path: "pages/a.md".into(),
            length: 3,
            digest: "a".repeat(64),
        }];
        let mut changed = first.clone();
        changed[0].digest = "b".repeat(64);
        assert_ne!(aggregate_digest(&first), aggregate_digest(&changed));
        assert_eq!(aggregate_digest(&first), aggregate_digest(&first));
    }
}
